drivers/counter/counter_ll_stm32_timer.c

```diff
@@ -0,0 +1,623 @@
+
+		cb(dev, id, cc_val, chdata->user_data);
+	}
+}
+
+static const struct counter_driver_api counter_stm32_driver_api = {
+	.start = counter_stm32_start,
+	.stop = counter_stm32_stop,
+	.get_value = counter_stm32_get_value,
+	.set_alarm = counter_stm32_set_alarm,
+	.cancel_alarm = counter_stm32_cancel_alarm,
+	.set_top_value = counter_stm32_set_top_value,
+	.get_pending_int = counter_stm32_get_pending_int,
+	.get_top_value = counter_stm32_get_top_value,
+	.get_guard_period = counter_stm32_get_guard_period,
+	.set_guard_period = counter_stm32_set_guard_period,
+};
+
+
+#define TIM_IRQ_HANDLE_CC(timx, cc)					       \
+	do {								       \
+		bool hw_irq = LL_TIM_IsActiveFlag_CC##cc(timer) &&	       \
+			      LL_TIM_IsEnabledIT_CC##cc(timer);		       \
+		if (hw_irq || (DEV_DATA(dev)->cc_int_pending & BIT(cc - 1))) { \
+			if (hw_irq) {					       \
+				LL_TIM_ClearFlag_CC##cc(timer);		       \
+			}						       \
+			counter_stm32_alarm_irq_handle(dev, cc - 1);	       \
+		}							       \
+	} while (0)
+
+void counter_stm32_irq_handler(const struct device *dev)
+{
+	TIM_TypeDef *timer = DEV_CFG(dev)->timer;
+
+	/* Capture compare events */
+	switch (counter_get_num_of_channels(dev)) {
+	case 4:
+		TIM_IRQ_HANDLE_CC(timer, 4);
+	case 3:
+		TIM_IRQ_HANDLE_CC(timer, 3);
+	case 2:
+		TIM_IRQ_HANDLE_CC(timer, 2);
+	case 1:
+		TIM_IRQ_HANDLE_CC(timer, 1);
+	}
+
+	/* TIM Update event */
+	if (LL_TIM_IsActiveFlag_UPDATE(timer) && LL_TIM_IsEnabledIT_UPDATE(timer)) {
+		LL_TIM_ClearFlag_UPDATE(timer);
+		counter_stm32_top_irq_handle(dev);
+	}
+}
+
+#define TIMER(idx)              DT_PARENT(DT_DRV_INST(idx))
+
+#define TIMER_IRQ_CONNECT(idx)				 \
+	IRQ_DIRECT_CONNECT(DT_IRQN(TIMER(idx)),		 \
+			   DT_IRQ(TIMER(idx), priority), \
+			   counter##idx##_irq_handler,	 \
+			   IRQ_ZERO_LATENCY)
+
+/** TIMx instance from DT */
+#define TIM(idx) ((TIM_TypeDef *)DT_REG_ADDR(TIMER(idx)))
+
+#define TIMER_CLK(idx)					\
+	{						\
+		.bus = DT_CLOCKS_CELL(TIMER(idx), bus),	\
+		.enr = DT_CLOCKS_CELL(TIMER(idx), bits)	\
+	}
+
+#define COUNTER_DEVICE_INIT(idx)						  \
+	BUILD_ASSERT(DT_INST_PROP(idx, st_prescaler) <=				  \
+		     0xFFFF,							  \
+		     "TIMER counter prescaler out of range");			  \
+	BUILD_ASSERT(NUM_CH(TIM(idx)) <= TIMER_MAX_CH,				  \
+		     "TIMER too many channels");				  \
+										  \
+	static struct counter_stm32_data counter##idx##_data;			  \
+	static struct counter_stm32_ch_data counter##idx##_ch_data[TIMER_MAX_CH]; \
+										  \
+	ISR_DIRECT_DECLARE(counter##idx##_irq_handler)				  \
+	{									  \
+		counter_stm32_irq_handler(DEVICE_DT_GET(DT_DRV_INST(idx)));	  \
+		return 0;							  \
+	}									  \
+										  \
+	static int counter##idx##_init(const struct device *dev)		  \
+	{									  \
+		TIMER_IRQ_CONNECT(idx);						  \
+		return counter_stm32_init_timer(dev);				  \
+	}									  \
+										  \
+	LOG_INSTANCE_REGISTER(LOG_MODULE_NAME, idx, CONFIG_COUNTER_LOG_LEVEL);	  \
```

Comment:
is it worth having a logger per-instance?

Reply:
I suppose not. Fixed.